File: Data Structure/C++ Programming Language/SinglyLinearLL.cpp
// Singly Circular Linked List

#include<iostream>
using namespace std;

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

class SinglyLinearLL
{
    private:
        PNODE first;
        int iCount;

    public:
        SinglyLinearLL()
        {
            cout<<"Inside Constructor of Singly LL\n";

            this->first = NULL;
            this->iCount = 0;
        }

        void InsertFirst(int no)
        {
            PNODE newn = NULL;

            newn = new NODE;

            newn->data = no;
            newn->next = NULL;

            if(first == NULL)
            {
                first = newn;
            }
            else
            {
                newn->next = first;
                first = newn;
            }

            iCount++;
        }

        void InsertLast(int no)
        {
            PNODE newn = NULL;
            PNODE temp = NULL;

            newn = new NODE;

            newn->data = no;
            newn->next = NULL;

            if(first == NULL)
            {
                first = newn;
            }
            else
            {
                temp = first;
                
                while((temp->next) != NULL)
                {
                    temp = temp->next;
                }
                temp->next = newn;
            }

            iCount++;
        }

        void InsertAtPos(int no, int pos)
        {
            PNODE newn = NULL;
            PNODE temp = NULL;

            int i = 0;

            if(pos < 1 && pos > (iCount + 1))
            {
                cout<<"Invalid Position\n";
                return;
            }

            if(pos == 1)
            {
                InsertFirst(no);
            }
            else if(pos == (iCount + 1))
            {
                InsertLast(no);
            }
            else
            {
                newn = new NODE;

                newn->data = no;
                newn->next = NULL;

                temp = first;

                for(i = 1; i < (pos - 1); i++)
                {
                    temp = temp->next;
                }

                newn->next = temp->next;
                temp->next = newn;

                iCount++;
            }
        }

        void DeleteFirst()
        {
            PNODE temp = NULL;

            if(first == NULL)
            {
                cout<<"Linked List is Empty";
                return;
            }
            else if(first->next == NULL)
            {
                delete(first);

                first = NULL;
            }
            else
            {
                temp = first;
                first = first->next;

                delete(temp);
            }

            iCount--;
        }

        void DeleteLast()
        {
            PNODE temp = NULL;

            if(first == NULL)
            {
                cout<<"Linked List is Empty";
                return;
            }
            else if(first->next == NULL)
            {
                delete(first);

                first = NULL;
            }
            else
            {
                temp = first;

                while(temp->next->next != NULL)
                {
                    temp = temp->next;
                }

                delete(temp->next);

                temp->next = NULL;
            }

            iCount--;
        }

        void DeleteAtPos(int pos)
        {
            PNODE temp = NULL;
            PNODE target = NULL;

            int i = 0;

            if(pos < 1 && pos > iCount)
            {
                cout<<"Invalid Position\n";
                return;
            }

            if(pos == 1)
            {
                DeleteFirst();
            }
            else if(pos == iCount)
            {
                DeleteLast();
            }
            else
            {
                temp = first;

                for(i = 1; i < (pos - 1); i++)
                {
                    temp = temp->next;
                }
                target = temp->next;
                
                temp->next = target->next;
                
                delete(target);

                iCount--;
            }
        }

        int Count()
        {
            return iCount;
        }

        void Display()
        {
            PNODE temp = NULL;

            if(first == NULL)
            {
                cout<<"Linked List is Empty";
                return;
            }

            temp = first;

            while(temp != NULL)
            {
                cout<<" | "<<temp->data<<" | -> ";
                temp = temp->next;
            }
            cout<<"NULL\n";
        }
};

int main()
{
    SinglyLinearLL sobj;

    int iRet = 0;

    sobj.InsertFirst(51);
    sobj.InsertFirst(21);
    sobj.InsertFirst(11);

    sobj.InsertLast(101);
    sobj.InsertLast(111);
    sobj.InsertLast(121);

    sobj.Display();

    iRet = sobj.Count();

    cout<<"Number of Elements in the LL are : "<<iRet<<"\n";

    // sobj.DeleteFirst();
    // sobj.Display();
    // iRet = sobj.Count();
    // cout<<"Number of Elements in the LL after Delete-First are : "<<iRet<<"\n";
    // sobj.DeleteLast();
    // sobj.Display();
    // iRet = sobj.Count();
    // cout<<"Number of Elements in the LL after Delete-Last are : "<<iRet<<"\n";

    sobj.InsertAtPos(105,5);

    sobj.Display();

    iRet = sobj.Count();

    cout<<"Number of Elements in the LL after Insert-At-Pos are : "<<iRet<<"\n";

    sobj.DeleteAtPos(5);

    sobj.Display();

    iRet = sobj.Count();

    cout<<"Number of Elements in the LL after Delete-At-Pos are : "<<iRet<<"\n";

    return 0;
}

File: Data Structure/Java Programming Language/SinglyLinearLL.java
// Singly Linear Linked List

class node
{
    public int data;
    public node next;                   // Self Referancial Member
}

class SinglyLL
{
    private node first;
    private int iCount;

    public SinglyLL()
    {
        System.out.println("Inside Constructor");
        first = null;
        iCount = 0;
    }

    public void InsertFirst(int no)
    {
        node newn = null;

        newn = new node();

        newn.data = no;
        newn.next = null;

        if(first == null)               // LL is Empty
        {
            first = newn;
        }
        else                            // LL contains atleast 1 node
        {
            newn.next = first;
            first = newn;
        }
        iCount++;
    }

    public void InsertLast(int no)
    {
        node newn = null;
        node temp = null;

        newn = new node();

        newn.data = no;
        newn.next = null;

        if(first == null)               // LL is Empty
        {
            first = newn;
        }
        else                            // LL contains atleast 1 node
        {
            temp = first;

            while(temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
        }
        iCount++;
    }

    public void Display()
    {
        node temp = null;

        temp = first;

        while(temp != null)
        {
            System.out.print("| "+temp.data+" | -> ");
            temp = temp.next;
        }

        System.out.println("null");
    }

    public int Count()
    {
        return iCount;
    }

    public void DeleteFirst()
    {
        if(first == null)                   // LL is Empty
        {
            return;
        }
        else if(first.next == null)         // LL contains atleast 1 node
        {
            first = null;
        }
        else                                // LL contains more than 1 node
        {
            first = first.next;
        }
        iCount--;
    }

    public void DeleteLast()
    {
        node temp = null;

        if(first == null)
        {
            return;
        }
        else
        {
            temp = first;

            while(temp.next.next != null)
            {
                temp = temp.next;
            }
            temp.next = null;
        }
        iCount--;
    }

    public void InsertAtPos(int no, int pos)
    {
        node newn = null;
        node temp = null;
        int iCnt = 0;

        if(pos < 1 || pos > iCount + 1)
        {
            System.out.println("Invalid Position");
            return;
        }

        if(pos == 1)
        {
            InsertFirst(no);
        }
        else if(pos == iCount + 1)
        {
            InsertLast(no);
        }
        else
        {
            newn = new node();

            newn.data = no;
            newn.next = null;

            temp = first;

            for(iCnt = 1; iCnt < (pos - 1); iCnt++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            temp.next = newn;

            iCount++;
        }
    }

    public void DeleteAtPos(int pos)
    {
        node temp = null;
        int iCnt = 0;

        if(pos < 1 || pos > iCount)
        {
            System.out.println("Invalid Position");
            return;
        }

        if(pos == 1)
        {
            DeleteFirst();
        }
        else if(pos == iCount)
        {
            DeleteLast();
        }
        else
        {
            temp = first;

            for(iCnt = 1; iCnt < (pos - 1); iCnt++)
            {
                temp = temp.next;
            }

            temp.next = temp.next.next;

            iCount--;
        }
    }
}

class SinglyLinearLL
{
    public static void main(String A[])
    {
        SinglyLL sobj = new SinglyLL();
        int iRet = 0;

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);

        sobj.InsertLast(101);
        sobj.InsertLast(111);
        sobj.InsertLast(121);

        sobj.Display();   

        iRet = sobj.Count();

        System.out.println("Number of Elements in the Linked List are : "+iRet);

        sobj.InsertAtPos(105,5);                     ///////////////////////////////////////

        sobj.Display();   

        iRet = sobj.Count();

        System.out.println("Number of Elements in the Linked List after Insert-at-Pos are : "+iRet);

        sobj.DeleteAtPos(5);

        sobj.Display();   

        iRet = sobj.Count();

        System.out.println("Number of Elements in the Linked List after Delete-at-Pos are : "+iRet);
    }
}